processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java

```diff
@@ -658,37 +671,42 @@ public DateTime getMaxIngestedEventTime()
     return maxIngestedEventTime;
   }
 
-  class DimensionHolder
+  private class DimensionHolder
   {
-    private final Map<String, DimDim> dimensions;
+    private DimDim[] dimensions;
 
-    DimensionHolder()
+    private DimensionHolder(int numDimension)
     {
-      dimensions = Maps.newConcurrentMap();
+      dimensions = new DimDim[Math.max(32, numDimension)];
```

Comment:
> Why is there a minimum of 32 entries created?

It would be any number which is >=1. Just choosed blindly. 

> also, the previous data structure was a concurrent map, how is locking done over this array?

Put() method is always call in lock (`dimensionOrder`), but missed lock for get(). I'll add one for that.